File: src/rac/services/review.py
"""Repository review — `rac review` (v0.7.9).

``build_review`` aggregates RAC's existing repository intelligence into one
prioritized, actionable report (REQ-Repository-Review-Mode): what needs
attention in this repository, ordered by impact, with a deterministic
suggested action per finding.

The service composes ``build_portfolio_summary`` — it implements no analysis
of its own ("Review Mode shall not duplicate existing RAC command logic";
ADR-015: consumers render, never analyze). Duplicate-identifier detection
remains with ``rac relationships --validate``, which performs the repo-level
identifier index walk.

Priority order (REQ-Repository-Review-Mode, "Repository Health Summary"):

1. Invalid artifacts (validation errors)
2. Broken relationships (unresolved, ambiguous, or self references)
3. Missing required information (unrecognized artifacts — no schema matched)
4. Missing recommended information (recommended sections absent)
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from pathlib import Path

from rac.core.corpus import CorpusCache

from .drift import CODE_SUSPECT_ARTIFACT, drift_problem, suspect_drift
from .portfolio import (
    ATTENTION_BROKEN_RELATIONSHIP,
    ATTENTION_INVALID,
    ATTENTION_MISSING_RECOMMENDED,
    PortfolioSummary,
    build_portfolio_summary,
)
from .recency import artifact_recency

# Stable priority levels and the unknown-artifact code (JSON contract, ADR-007).
PRIORITY_INVALID_ARTIFACT = 1
PRIORITY_BROKEN_RELATIONSHIP = 2
PRIORITY_UNKNOWN_ARTIFACT = 3
PRIORITY_MISSING_RECOMMENDED = 4
# Write-cadence nudge (v0.13.3): below every other finding, never fails review.
PRIORITY_STALE_CORPUS = 5
# Suspect-artifact drift (freshness-and-drift phase 1): a git-native advisory that
# sits beside the cadence nudge — advisory only, never fails review (REQ-002).
PRIORITY_SUSPECT_DRIFT = 6

REVIEW_UNKNOWN_ARTIFACT = "unknown-artifact"
REVIEW_STALE_CORPUS = "stale-corpus"
# The stable drift code is owned by `rac.services.drift`; re-exported here under
# the review namespace so both surfaces report the identical code (ADR-007).
REVIEW_SUSPECT_ARTIFACT = CODE_SUSPECT_ARTIFACT

# Default cadence window when `--stale-after` is given without a value: two
# weeks (v0.13.3).
DEFAULT_STALE_AFTER_DAYS = 14

# Attention code -> review priority for findings inherited from the portfolio.
_ATTENTION_PRIORITY = {
    ATTENTION_INVALID: PRIORITY_INVALID_ARTIFACT,
    ATTENTION_BROKEN_RELATIONSHIP: PRIORITY_BROKEN_RELATIONSHIP,
    ATTENTION_MISSING_RECOMMENDED: PRIORITY_MISSING_RECOMMENDED,
}

# "Why it matters" per finding code (v0.8.11) — Core owns the impact text so
# every consumer (JSON, CLI, Explorer) reads the same sentence. Moved here
# from the Explorer adapter; an unrecognized code gets the generic sentence,
# so the field is always present.
_GENERIC_IMPACT = "This finding affects repository quality."
_IMPACT = {
    ATTENTION_INVALID: "The artifact fails its schema, so tooling and validation cannot trust it.",
    ATTENTION_BROKEN_RELATIONSHIP: (
        "A declared reference does not resolve, leaving traceability incomplete."
    ),
    ATTENTION_MISSING_RECOMMENDED: (
        "Recommended sections are empty, weakening the artifact's completeness."
    ),
    REVIEW_UNKNOWN_ARTIFACT: "No schema matched, so required structure cannot be checked.",
    REVIEW_STALE_CORPUS: (
        "The write habit has stalled; product knowledge stops reflecting the work."
    ),
    REVIEW_SUSPECT_ARTIFACT: (
        "A referenced artifact changed after this one did, so the reference may be stale."
    ),
}


def impact_for(code: str) -> str:
    """The Core-owned impact sentence for a finding ``code``."""
    return _IMPACT.get(code, _GENERIC_IMPACT)


@dataclass
class ReviewIssue:
    """One prioritized finding with its deterministic next step."""

    priority: int  # 1 (highest impact) – 4
    severity: str  # "error" | "warning" | "info"
    path: str
    identifier: str  # artifact identifier or filename stem
    code: str
    message: str
    action: str  # a runnable command or concrete edit
    impact: str  # why it matters (v0.8.11; additive JSON field, ADR-007)

    def to_dict(self) -> dict:
        return {
            "priority": self.priority,
            "severity": self.severity,
            "path": self.path,
            "identifier": self.identifier,
            "code": self.code,
            "message": self.message,
            "action": self.action,
            "impact": self.impact,
        }


@dataclass
class ReviewReport:
    """Repository review result (v0.7.9).

    ``to_dict`` is the stable JSON contract (ADR-007); fields are additive and
    schema_version-gated. The inventory/validation/relationship/health blocks
    mirror the ``rac portfolio`` contract so consumers can share parsing.
    """

    directory: str
    recursive: bool
    portfolio: PortfolioSummary
    issues: list[ReviewIssue]

    @property
    def ok(self) -> bool:
        """True when nothing demands attention before work continues.

        Priority 1–2 findings (invalid artifacts, broken relationships) fail
        the review; priority 3–4 findings are advisory.
        """
        return not any(i.priority <= PRIORITY_BROKEN_RELATIONSHIP for i in self.issues)

    @property
    def actions(self) -> list[str]:
        """Deduplicated suggested actions in priority order."""
        seen: set[str] = set()
        ordered: list[str] = []
        for issue in self.issues:
            if issue.action not in seen:
                seen.add(issue.action)
                ordered.append(issue.action)
        return ordered

    def to_dict(self) -> dict:
        p = self.portfolio
        return {
            "schema_version": "1",
            "directory": self.directory,
            "recursive": self.recursive,
            "ok": self.ok,
            # Additive in v0.13.1 (ADR-007): a day-one empty-corpus marker.
            "empty": p.total_artifacts == 0,
            "artifacts": {
                "total": p.total_artifacts,
                "by_type": p.by_type,
                "unknown_paths": p.unknown_paths,
            },
            "validation": {
                "valid": p.valid_artifacts,
                "invalid": p.invalid_artifacts,
            },
            "relationships": {
                "total": p.relationships.total,
                "valid": p.relationships.valid,
                "broken": p.relationships.broken,
                "orphaned": p.relationships.orphaned,
                "coverage": p.relationships.coverage,
            },
            "health": {
                "score": p.health_score,
            },
            "issues": [i.to_dict() for i in self.issues],
            "actions": self.actions,
        }


def build_review(
    directory: str,
    recursive: bool = True,
    *,
    stale_after_days: int | None = None,
    now: datetime | None = None,
) -> ReviewReport:
    """Review ``directory`` and return the prioritized repository report.

    With ``stale_after_days`` set, an advisory write-cadence finding is added
    when the corpus has had no new or updated artifact within the window
    (v0.13.3). It is informational and never changes the review's exit status.
    ``now`` is injectable for deterministic tests.
    """
    portfolio = build_portfolio_summary(directory, recursive=recursive)
    report = review_from_portfolio(directory, portfolio, recursive=recursive)
    advisories: list[ReviewIssue] = list(_drift_findings(directory, recursive))
    if stale_after_days is not None:
        finding = _cadence_finding(directory, recursive, stale_after_days, now=now)
        if finding is not None:
            advisories.append(finding)
    if advisories:
        report.issues.extend(advisories)
        report.issues.sort(key=lambda i: (i.priority, i.path, i.code))
    return report


def _drift_findings(directory: str, recursive: bool) -> list[ReviewIssue]:
    """Suspect-artifact drift advisories, beside the cadence nudge (REQ-002).

    Surfaces the same git-native signal ``rac doctor`` reports, through review's
    advisory channel: one finding per referrer whose resolved target changed more
    recently. Advisory only (priority below every blocking finding), so it never
    changes the review verdict; empty outside git or with no drift (REQ-005).
    """
    entries = CorpusCache().collect(directory, recursive=recursive)
    issues: list[ReviewIssue] = []
    for record in suspect_drift(directory, entries):
        issues.append(
            ReviewIssue(
                priority=PRIORITY_SUSPECT_DRIFT,
                severity="warning",
                path=record.source_path,
                identifier=Path(record.source_path).stem,
                code=REVIEW_SUSPECT_ARTIFACT,
                message=drift_problem(record),
                action="Run: rac doctor " + directory,
                impact=impact_for(REVIEW_SUSPECT_ARTIFACT),
            )
        )
    return issues


def _cadence_finding(
    directory: str,
    recursive: bool,
    window_days: int,
    *,
    now: datetime | None = None,
) -> ReviewIssue | None:
    """The write-cadence nudge, or ``None`` when it should not fire.

    Fires only when recency is known (inside git, with committed artifacts)
    and the newest artifact is older than ``window_days``. An empty corpus or
    unknown recency is suppressed — the v0.13.1 empty-corpus hint covers the
    day-one case, and a nudge on missing data would be noise.
    """
    recency = artifact_recency(directory, recursive=recursive)
    most_recent = recency.most_recent
    if most_recent is None:
        return None
    moment = now or datetime.now(UTC)
    age = moment - most_recent
    if age <= timedelta(days=window_days):
        return None
    return ReviewIssue(
        priority=PRIORITY_STALE_CORPUS,
        severity="info",
        path=directory,
        identifier="corpus",
        code=REVIEW_STALE_CORPUS,
        message=(
            f"No product knowledge recorded in the last {window_days} days "
            f"(newest artifact is {age.days} days old)."
        ),
        action="Run: rac new decision rac/decisions/<name>.md",
        impact=impact_for(REVIEW_STALE_CORPUS),
    )


def review_from_portfolio(
    directory: str, portfolio: PortfolioSummary, recursive: bool = True
) -> ReviewReport:
    """Build the review from an already-computed portfolio (v0.8.3).

    Same result as :func:`build_review`; the seam lets a consumer holding a
    loaded repository model (Explorer) reuse Core's review logic without a
    second walk (ADR-015: the recommendation logic stays here, not in Explorer).
    """
    issues: list[ReviewIssue] = []

    # Findings the portfolio already computed, re-ranked by review priority
    # and paired with a deterministic next step.
    for item in portfolio.attention:
        priority = _ATTENTION_PRIORITY.get(item.code)
        if priority is None:  # future attention codes: surface, lowest priority
            priority = PRIORITY_MISSING_RECOMMENDED
        if item.code == ATTENTION_INVALID:
            action = f"Run: rac validate {item.path}"
        elif item.code == ATTENTION_BROKEN_RELATIONSHIP:
            action = f"Run: rac relationships {directory} --validate"
        else:
            action = f"Run: rac improve {item.path} --template"
        issues.append(
            ReviewIssue(
                priority=priority,
                severity=item.severity,
                path=item.path,
                identifier=item.identifier,
                code=item.code,
                message=item.message,
                action=action,
                impact=impact_for(item.code),
            )
        )

    # Unrecognized artifacts: no schema matched, so required information is
    # missing by definition (priority 3). Advisory — Unknown is a valid outcome.
    for path in portfolio.unknown_paths:
        issues.append(
            ReviewIssue(
                priority=PRIORITY_UNKNOWN_ARTIFACT,
                severity="info",
                path=path,
                identifier=Path(path).stem,
                code=REVIEW_UNKNOWN_ARTIFACT,
                message="No artifact schema matched this document.",
                action=f"Run: rac inspect {path} (see rac schema --list)",
                impact=impact_for(REVIEW_UNKNOWN_ARTIFACT),
            )
        )

    # Deterministic order: impact first, then path, then code.
    issues.sort(key=lambda i: (i.priority, i.path, i.code))

    return ReviewReport(
        directory=directory,
        recursive=recursive,
        portfolio=portfolio,
        issues=issues,
    )

File: rust/rac-engine/src/review.rs
//! Repository review (`rac.services.review`) — the prioritized, actionable
//! report `rac review` renders. Composes the portfolio summary, git-native
//! drift advisories, and an optional write-cadence nudge.

use std::path::{Path, PathBuf};

use crate::gitinfo;
use crate::portfolio::{
    portfolio_from_corpus, AttentionItem, PortfolioSummary, ATTENTION_BROKEN_RELATIONSHIP,
    ATTENTION_INVALID, ATTENTION_MISSING_RECOMMENDED,
};
use crate::relationships::{corpus_items, relationships_from_corpus, CorpusItem};

pub const PRIORITY_INVALID_ARTIFACT: i64 = 1;
pub const PRIORITY_BROKEN_RELATIONSHIP: i64 = 2;
pub const PRIORITY_UNKNOWN_ARTIFACT: i64 = 3;
pub const PRIORITY_MISSING_RECOMMENDED: i64 = 4;
pub const PRIORITY_STALE_CORPUS: i64 = 5;
pub const PRIORITY_SUSPECT_DRIFT: i64 = 6;

pub const REVIEW_UNKNOWN_ARTIFACT: &str = "unknown-artifact";
pub const REVIEW_STALE_CORPUS: &str = "stale-corpus";
pub const REVIEW_SUSPECT_ARTIFACT: &str = "suspect-artifact";

const GENERIC_IMPACT: &str = "This finding affects repository quality.";

fn impact_for(code: &str) -> &'static str {
    match code {
        ATTENTION_INVALID => {
            "The artifact fails its schema, so tooling and validation cannot trust it."
        }
        ATTENTION_BROKEN_RELATIONSHIP => {
            "A declared reference does not resolve, leaving traceability incomplete."
        }
        ATTENTION_MISSING_RECOMMENDED => {
            "Recommended sections are empty, weakening the artifact's completeness."
        }
        REVIEW_UNKNOWN_ARTIFACT => {
            "No schema matched, so required structure cannot be checked."
        }
        REVIEW_STALE_CORPUS => {
            "The write habit has stalled; product knowledge stops reflecting the work."
        }
        REVIEW_SUSPECT_ARTIFACT => {
            "A referenced artifact changed after this one did, so the reference may be stale."
        }
        _ => GENERIC_IMPACT,
    }
}

#[derive(Debug, Clone)]
pub struct ReviewIssue {
    pub priority: i64,
    pub severity: String,
    pub path: String,
    pub identifier: String,
    pub code: String,
    pub message: String,
    pub action: String,
    pub impact: String,
}

pub struct ReviewReport {
    pub directory: String,
    pub recursive: bool,
    pub portfolio: PortfolioSummary,
    pub issues: Vec<ReviewIssue>,
}

impl ReviewReport {
    pub fn ok(&self) -> bool {
        !self
            .issues
            .iter()
            .any(|i| i.priority <= PRIORITY_BROKEN_RELATIONSHIP)
    }

    /// Deduplicated suggested actions in issue (priority) order.
    pub fn actions(&self) -> Vec<String> {
        let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();
        let mut ordered = Vec::new();
        for issue in &self.issues {
            if seen.insert(issue.action.clone()) {
                ordered.push(issue.action.clone());
            }
        }
        ordered
    }
}

fn attention_priority(code: &str) -> i64 {
    match code {
        ATTENTION_INVALID => PRIORITY_INVALID_ARTIFACT,
        ATTENTION_BROKEN_RELATIONSHIP => PRIORITY_BROKEN_RELATIONSHIP,
        ATTENTION_MISSING_RECOMMENDED => PRIORITY_MISSING_RECOMMENDED,
        _ => PRIORITY_MISSING_RECOMMENDED,
    }
}

fn sort_issues(issues: &mut [ReviewIssue]) {
    issues.sort_by(|a, b| {
        a.priority
            .cmp(&b.priority)
            .then(a.path.cmp(&b.path))
            .then(a.code.cmp(&b.code))
    });
}

pub fn review_from_portfolio(
    directory: &str,
    portfolio: PortfolioSummary,
    recursive: bool,
) -> ReviewReport {
    let mut issues: Vec<ReviewIssue> = Vec::new();

    for item in &portfolio.attention {
        let AttentionItem {
            path,
            identifier,
            severity,
            code,
            message,
        } = item;
        let priority = attention_priority(code);
        let action = if code == ATTENTION_INVALID {
            format!("Run: rac validate {path}")
        } else if code == ATTENTION_BROKEN_RELATIONSHIP {
            format!("Run: rac relationships {directory} --validate")
        } else {
            format!("Run: rac improve {path} --template")
        };
        issues.push(ReviewIssue {
            priority,
            severity: severity.clone(),
            path: path.clone(),
            identifier: identifier.clone(),
            code: code.clone(),
            message: message.clone(),
            action,
            impact: impact_for(code).to_string(),
        });
    }

    for path in &portfolio.unknown_paths {
        issues.push(ReviewIssue {
            priority: PRIORITY_UNKNOWN_ARTIFACT,
            severity: "info".to_string(),
            path: path.clone(),
            identifier: crate::identity::path_stem(path),
            code: REVIEW_UNKNOWN_ARTIFACT.to_string(),
            message: "No artifact schema matched this document.".to_string(),
            action: format!("Run: rac inspect {path} (see rac schema --list)"),
            impact: impact_for(REVIEW_UNKNOWN_ARTIFACT).to_string(),
        });
    }

    sort_issues(&mut issues);

    ReviewReport {
        directory: directory.to_string(),
        recursive,
        portfolio,
        issues,
    }
}

/// `build_review(directory, recursive, stale_after_days)`.
pub fn build_review(
    directory: &str,
    recursive: bool,
    stale_after_days: Option<i64>,
) -> ReviewReport {
    let items = corpus_items(directory, recursive);
    let portfolio = portfolio_from_corpus(directory, &items, recursive);
    let mut report = review_from_portfolio(directory, portfolio, recursive);

    let mut advisories: Vec<ReviewIssue> = drift_findings(directory, &items);
    if let Some(window) = stale_after_days {
        if let Some(finding) = cadence_finding(directory, &items, window) {
            advisories.push(finding);
        }
    }
    if !advisories.is_empty() {
        report.issues.extend(advisories);
        sort_issues(&mut report.issues);
    }
    report
}

// --- git-native drift --------------------------------------------------------

pub(crate) struct DriftRecord {
    pub(crate) source_path: String,
    pub(crate) target_path: String,
    pub(crate) target_ref: String,
    pub(crate) source_committed: String,
    pub(crate) target_committed: String,
}

/// `suspect_drift(directory, entries)` — resolved edges whose target was
/// committed strictly after the referrer. Deduped per `(source, target)`,
/// sorted by `(source_path, target_path)`.
pub(crate) fn suspect_drift(directory: &str, items: &[CorpusItem]) -> Vec<DriftRecord> {
    let resolved: Vec<crate::relationships::Relationship> = relationships_from_corpus(items)
        .into_iter()
        .filter(|r| r.resolved_path.is_some())
        .collect();
    if resolved.is_empty() {
        return Vec::new();
    }

    let mut involved: Vec<PathBuf> = Vec::new();
    let mut seen_paths: std::collections::HashSet<String> = std::collections::HashSet::new();
    for rel in &resolved {
        for p in [&rel.source_path, rel.resolved_path.as_ref().unwrap()] {
            if seen_paths.insert(p.clone()) {
                involved.push(PathBuf::from(p));
            }
        }
    }
    let committed_pairs = gitinfo::last_committed_for_paths(Path::new(directory), &involved);
    let committed: std::collections::HashMap<String, Option<String>> = committed_pairs
        .into_iter()
        .map(|(p, v)| (p.to_string_lossy().into_owned(), v))
        .collect();

    let mut records: Vec<DriftRecord> = Vec::new();
    let mut seen: std::collections::HashSet<(String, String)> = std::collections::HashSet::new();
    for rel in &resolved {
        let target_path = rel.resolved_path.clone().unwrap();
        let source_when = committed.get(&rel.source_path).and_then(|v| v.clone());
        let target_when = committed.get(&target_path).and_then(|v| v.clone());
        let (source_when, target_when) = match (source_when, target_when) {
            (Some(s), Some(t)) => (s, t),
            _ => continue,
        };
        let source_epoch = gitinfo::parse_iso8601_epoch(&source_when);
        let target_epoch = gitinfo::parse_iso8601_epoch(&target_when);
        // target newer than source (strictly) => suspect.
        match (source_epoch, target_epoch) {
            (Some(se), Some(te)) if te > se => {}
            _ => continue,
        }
        let key = (rel.source_path.clone(), target_path.clone());
        if !seen.insert(key) {
            continue;
        }
        records.push(DriftRecord {
            source_path: rel.source_path.clone(),
            target_path: target_path.clone(),
            target_ref: rel.target.clone(),
            source_committed: source_when,
            target_committed: target_when,
        });
    }
    records.sort_by(|a, b| {
        a.source_path
            .cmp(&b.source_path)
            .then(a.target_path.cmp(&b.target_path))
    });
    records
}

fn drift_findings(directory: &str, items: &[CorpusItem]) -> Vec<ReviewIssue> {
    suspect_drift(directory, items)
        .into_iter()
        .map(|record| ReviewIssue {
            priority: PRIORITY_SUSPECT_DRIFT,
            severity: "warning".to_string(),
            path: record.source_path.clone(),
            identifier: crate::identity::path_stem(&record.source_path),
            code: REVIEW_SUSPECT_ARTIFACT.to_string(),
            message: drift_problem(&record),
            action: format!("Run: rac doctor {directory}"),
            impact: impact_for(REVIEW_SUSPECT_ARTIFACT).to_string(),
        })
        .collect()
}

pub(crate) fn drift_problem(record: &DriftRecord) -> String {
    format!(
        "references {} which changed more recently (target last committed {}, this artifact {}) — review recommended",
        record.target_ref,
        gitinfo::isoformat_roundtrip(&record.target_committed),
        gitinfo::isoformat_roundtrip(&record.source_committed),
    )
}

// --- write-cadence nudge -----------------------------------------------------

fn cadence_finding(
    directory: &str,
    items: &[CorpusItem],
    window_days: i64,
) -> Option<ReviewIssue> {
    // most_recent = newest last-committed across recognised (non-unknown) artifacts.
    let recognised: Vec<PathBuf> = items
        .iter()
        .filter(|i| i.spec.is_some())
        .map(|i| PathBuf::from(&i.path))
        .collect();
    if recognised.is_empty() {
        return None;
    }
    let committed = gitinfo::last_committed_for_paths(Path::new(directory), &recognised);
    let mut most_recent_epoch: Option<i64> = None;
    let mut most_recent_stamp: Option<String> = None;
    for (_, stamp) in committed {
        if let Some(stamp) = stamp {
            if let Some(epoch) = gitinfo::parse_iso8601_epoch(&stamp) {
                if most_recent_epoch.map(|e| epoch > e).unwrap_or(true) {
                    most_recent_epoch = Some(epoch);
                    most_recent_stamp = Some(stamp);
                }
            }
        }
    }
    let (most_recent_epoch, _) = (most_recent_epoch?, most_recent_stamp);

    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0);
    // age.days = floor((now - most_recent)/86400); age <= window => suppress.
    let delta = now - most_recent_epoch;
    let age_days = delta.div_euclid(86_400);
    if age_days <= window_days {
        return None;
    }
    Some(ReviewIssue {
        priority: PRIORITY_STALE_CORPUS,
        severity: "info".to_string(),
        path: directory.to_string(),
        identifier: "corpus".to_string(),
        code: REVIEW_STALE_CORPUS.to_string(),
        message: format!(
            "No product knowledge recorded in the last {window_days} days (newest artifact is {age_days} days old)."
        ),
        action: "Run: rac new decision rac/decisions/<name>.md".to_string(),
        impact: impact_for(REVIEW_STALE_CORPUS).to_string(),
    })
}
